tests/Unit/BaseCollectionTest.php

```diff
@@ -132,4 +134,30 @@ public function testJsonSerialize(): void
             $this->fakeBaseCollection->jsonSerialize()
         );
     }
+
+    public function testFilter(): void
+    {
+        $collection = new Collection([
+            new Element('need'),
+            new Element('need'),
+            new Element('not need'),
+            new Element('not need'),
+            new Element('need'),
+        ]);
+
+        $filtered = $collection->filter(function (Element $element): bool {
+            return $element->value === 'need';
+        });
+
+        $this->assertNotSame($collection, $filtered);
+
+        $this->assertArrayHasKey(0, $filtered);
+        $this->assertEquals('need', $filtered[0]->value);
+        $this->assertArrayHasKey(1, $filtered);
+        $this->assertEquals('need', $filtered[1]->value);
+        $this->assertArrayNotHasKey(2, $filtered);
+        $this->assertArrayNotHasKey(3, $filtered);
+        $this->assertArrayHasKey(4, $filtered);
+        $this->assertEquals('need', $filtered[4]->value);
```

Comment:
Why not to use `assertEquals()` and pass expected array there?

Reply:
Because return value is not array. It is ArrayObject. AssertEquals of arrays will not pass by types of values